documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,888 @@
+
+For example:
+
+- We can implement the VNET routing by creating a routing type named `vnet` with only 1 action `staticencap`, which takes the next hop information from the mapping table.
+- We can also add additional hop to tunnel the traffic to an firewall or network virtual appliance first, which is known as [UDR](https://learn.microsoft.com/en-us/azure/virtual-network/virtual-networks-udr-overview), by creating another routing type named `vnetfw` with a `staticencap` action and an extra `tunnel` actions.
+- We can even use `tunnel` action with multiple destinations as ECMP group to implement a simple load balancer!
+
+This combination of routing actions is very flexible and powerful, and it enables us to implement any network function we want.
+
+### 5.9. Matching stages and metadata publishing
+
+#### 5.9.1. Matching stage
+
+Matching stage is the one of the core part of the DASH-SAI pipeline and the components that gives the pipeline flexibility.
+
+In DASH-SAI pipeline, a matching stage is a basic building block for packet matching and metadata publishing. And the DASH-SAI pipeline basically works as matching and metadata publishing until the final routing actions are executed. Currently, we support the following types of matching stages:
+
+| Stage type | Match Type | Match Fields | Metadata Behavior |
+| ---------- | ---------- | ------------ | -------- |
+| Routing | LPM | Source IP or Destination IP | Publish metadata from matched routing entry |
+| IP Mapping | Exact Match | Source IP or Destination IP | Publish metadata from matched VNET and VNET mapping entry |
+| Port Mapping | Range Match | Source Port + Destination Port | Publish metadata from matched port mapping entry |
+
+For more on the metadata publishing, please refer to the metadata publishing section below.
+
+#### 5.9.2. Pipeline profile and stage connections
+
+Ideally, when DASH initializes or whenever we create a new pipeline, by simply creating multiple numbers of different types of matching stages and connecting them in different ways, we can easily implement different pipeline and network functions.
+
+| Stage 0 | Stage 1 | Stage 2 | Stage 3 | Stage 4 |
+| ------- | ------- | ------- | ------- | ------- |
+| Routing 0 | IP Mapping 0 | Port Mapping | X | X |
+| Routing 0 | Routing 1 | IP Mapping 0 | IP Mapping 1 | Port Mapping |
+| Routing 0 | IP Mapping 0 | Port Mapping | Routing 1 | IP Mapping 1 |
+
+However, in reality, it might make the pipeline hard to implement, model, debug and validate or test at this moment. For example, changing matching fields, matching type and connection dynamically on pipeline creation is simply beyond the ability of P4, which is used by our behavior model today as well as many vendor's ASIC SDKs.
```

Comment:
Do we want to mention P4 annotations that may help P4 bmv2?

Reply:
p4 annotations will help generating the sai headers, but sadly, it won't be able to help much in this case. :(